src/Simplex/FileTransfer/Agent.hs

```diff
@@ -546,24 +563,44 @@ runXFTPSndWorker c srv Worker {doWork} = do
           any (\SndFileChunkReplica {replicaStatus} -> replicaStatus == SFRSUploaded) replicas
 
 deleteSndFileInternal :: AgentMonad m => AgentClient -> SndFileId -> m ()
-deleteSndFileInternal c sndFileEntityId = do
-  SndFile {sndFileId, prefixPath, status} <- withStore c $ \db -> getSndFileByEntityId db sndFileEntityId
-  if status == SFSComplete || status == SFSError
-    then do
-      forM_ prefixPath $ removePath <=< toFSFilePath
-      withStore' c (`deleteSndFile'` sndFileId)
-    else withStore' c (`updateSndFileDeleted` sndFileId)
+deleteSndFileInternal c sndFileEntityId = deleteSndFilesInternal c [sndFileEntityId]
+
+deleteSndFilesInternal :: AgentMonad m => AgentClient -> [SndFileId] -> m ()
+deleteSndFilesInternal c sndFileEntityIds = do
+  (_, sndFiles) <- partitionEithers <$> withStoreBatch c (\db -> map (getById db) sndFileEntityIds)
+  let pathsToRemove = filter (\SndFile {status} -> status == SFSComplete || status == SFSError) sndFiles
+  forM_ pathsToRemove $ \SndFile {prefixPath} ->
+    forM_ prefixPath (removePath <=< toFSFilePath) `catchAgentError` const (pure ())
+  void $ withStoreBatch' c (\db -> map (deleteFile db) sndFiles)
+  where
+    getById :: DB.Connection -> SndFileId -> IO (Either AgentErrorType SndFile)
+    getById db sndFileEntityId =
+      fmap (first storeError) $ runExceptT $ ExceptT $ getSndFileByEntityId db sndFileEntityId
+    deleteFile :: DB.Connection -> SndFile -> IO ()
+    deleteFile db SndFile {sndFileId, status} =
+      if status == SFSComplete || status == SFSError
+        then deleteSndFile' db sndFileId
+        else updateSndFileDeleted db sndFileId
 
 deleteSndFileRemote :: forall m. AgentMonad m => AgentClient -> UserId -> SndFileId -> ValidFileDescription 'FSender -> m ()
-deleteSndFileRemote c userId sndFileEntityId (ValidFileDescription FileDescription {chunks}) = do
-  deleteSndFileInternal c sndFileEntityId `catchAgentError` (notify c sndFileEntityId . SFERR)
-  forM_ chunks $ \ch -> deleteFileChunk ch `catchAgentError` (notify c sndFileEntityId . SFERR)
+deleteSndFileRemote c userId sndFileEntityId sfd = deleteSndFilesRemote c userId [(sndFileEntityId, sfd)]
+
+deleteSndFilesRemote :: forall m. AgentMonad m => AgentClient -> UserId -> [(SndFileId, ValidFileDescription 'FSender)] -> m ()
+deleteSndFilesRemote c userId sndFileIdsDescrs = do
+  deleteSndFilesInternal c (map fst sndFileIdsDescrs) `catchAgentError` (notify c "" . SFERR)
+  let chunks = concatMap (fdChunks . snd) sndFileIdsDescrs
+  void $ withStoreBatch' c (\db -> map (deleteChunk db) chunks)
+  let servers = S.fromList $ mapMaybe chunkServer chunks
+  forM_ servers $ \server -> void $ getXFTPDelWorker True c server
```

Comment:
this solves the same problem twice, in different ways, hiding what is happening - effectively in both cases you need the first replica